FprimeZephyrReference/Components/DetumbleManager/DetumbleManager.cpp

```diff
@@ -0,0 +1,432 @@
+
+    // Transition to COOLDOWN state
+    this->m_state = DetumbleState::COOLDOWN;
+}
+
+void DetumbleManager ::bdotTorqueAction() {
+    // Get magnetic field for dipole moment calculation
+    Fw::Success condition;
+    Drv::MagneticField magnetic_field = this->magneticFieldGet_out(0, condition);
+    if (condition != Fw::Success::SUCCESS) {
+        this->log_WARNING_LO_MagneticFieldRetrievalFailed();
+        return;
+    }
+    this->log_WARNING_LO_MagneticFieldRetrievalFailed_ThrottleClear();
+
+    // Get gain parameter
+    Fw::ParamValid isValid;
+    F64 gain = this->paramGet_GAIN(isValid);
+    this->tlmWrite_GainParam(gain);
+
+    // Get magnetometer sampling period
+    Fw::TimeIntervalValue sampling_period = this->magneticFieldSamplingPeriodGet_out(0, condition);
+    if (condition != Fw::Success::SUCCESS) {
+        this->log_WARNING_LO_MagneticFieldSamplingPeriodRetrievalFailed();
+        return;
+    }
+    this->log_WARNING_LO_MagneticFieldSamplingPeriodRetrievalFailed_ThrottleClear();
+
+    // Get dipole moment
+    errno = 0;  // Clear errno
+    std::array<double, 3> magnetic_field_array = {magnetic_field.get_x(), magnetic_field.get_y(),
+                                                  magnetic_field.get_z()};
+    std::chrono::microseconds sampling_period_us(sampling_period.get_useconds());
+    std::array<double, 3> dipole_moment =
+        this->m_bdot.getDipoleMoment(magnetic_field_array, magnetic_field.get_timestamp().get_seconds(),
+                                     magnetic_field.get_timestamp().get_useconds(), gain, sampling_period_us);
+
+    // Telemeter time between magnetic field readings
+    std::chrono::microseconds dt_readings = this->m_bdot.getTimeBetweenReadings();
+    this->tlmWrite_TimeBetweenMagneticFieldReadings(Fw::TimeIntervalValue(
+        static_cast<U32>(dt_readings.count() / 1000000), static_cast<U32>(dt_readings.count() % 1000000)));
+
+    // Check for errors in dipole moment computation
+    int rc = errno;
+    if (rc != 0) {
+        switch (rc) {
+            case EAGAIN:
+                return;
+            case EDOM:
+                this->log_WARNING_LO_MagneticFieldTooSmallForDipoleMoment();
+                return;
+            case EINVAL:
+                this->log_WARNING_LO_InvalidMagneticFieldReadingForDipoleMoment();
+                return;
+            default:
+                this->log_WARNING_LO_UnknownDipoleMomentComputationError();
+                return;
+        }
+    }
+    this->log_WARNING_LO_MagneticFieldTooSmallForDipoleMoment_ThrottleClear();
+    this->log_WARNING_LO_InvalidMagneticFieldReadingForDipoleMoment_ThrottleClear();
+    this->log_WARNING_LO_UnknownDipoleMomentComputationError_ThrottleClear();
+
+    // Perform torqueing action
+    this->startMagnetorquers(this->m_x_plus_magnetorquer.dipoleMomentToCurrent(dipole_moment[0]),
+                             this->m_x_minus_magnetorquer.dipoleMomentToCurrent(dipole_moment[0]),
```

Comment:
Review with Evan:
Does powering both magnitorquers in axis overcompensate the requested current? YES! Oops.